Wrappers/Python/cil/optimisation/algorithms/SPDHG.py

```diff
@@ -124,100 +247,118 @@ def set_up(self, f, g, operator, tau=None, sigma=None, \
+            precalculated list of norms of the operators 
         '''
         logging.info("{} setting up".format(self.__class__.__name__, ))
-                    
+
         # algorithmic parameters
         self.f = f
         self.g = g
         self.operator = operator
-        self.tau = tau
-        self.sigma = sigma
-        self.prob = prob
-        self.ndual_subsets = len(self.operator)
-        self.gamma = gamma
-        self.rho = .99
-        
-        if self.prob is None:
-            self.prob = [1/self.ndual_subsets] * self.ndual_subsets
-
-        
-        if self.sigma is None:
-            if norms is None:
-                # Compute norm of each sub-operator       
-                norms = [operator.get_item(i,0).norm() for i in range(self.ndual_subsets)]
-            self.norms = norms
-            self.sigma = [self.gamma * self.rho / ni for ni in norms] 
-        if self.tau is None:
-            self.tau = min( [ pi / ( si * ni**2 ) for pi, ni, si in zip(self.prob, norms, self.sigma)] ) 
-            self.tau *= (self.rho / self.gamma)
-
-        # initialize primal variable 
+        self.ndual_subsets = self.operator.shape[0]
+
+        if precalculated_norms is None:
+            # Compute norm of each sub-operator
+            self.norms = [self.operator.get_item(i, 0).norm()
+                          for i in range(self.ndual_subsets)]
+        else:
+            if len(precalculated_norms) == self.ndual_subsets:
+                if all(isinstance(x, Number) for x in precalculated_norms):
+                    if all(x > 0 for x in precalculated_norms):
+                        pass
+                    else:
+                        raise ValueError(
+                            "The norms of the operators should be positive")
+                else:
+                    raise ValueError(
+                        "The norms of the operators should be a Number")
+            else:
+                raise ValueError(
+                    "Please pass a list of floats to the precalculated norms with the same number of entries as number of operators")
+            self.norms = precalculated_norms
+
+        if sampler is None:
+            if self.prob_weights is None:
+                self.prob_weights = [1/self.ndual_subsets] * self.ndual_subsets
+            self.sampler = Sampler.randomWithReplacement(
+                self.ndual_subsets,  prob=self.prob_weights)
+        else:
+            if not isinstance(sampler, Sampler):
+                raise ValueError(
+                    "The sampler should be an instance of the CIL Sampler class")
+            self.sampler = sampler
+            if sampler.prob is None:
+                self.prob_weights = [1/self.ndual_subsets] * self.ndual_subsets
+            else:
+                self.prob_weights = sampler.prob
+
+        # might not want to do this until it is called (if computationally expensive)
+        self.set_step_sizes_custom()
```

Comment:
This seems to be used to set the default step sizes. Why do we call this `set_step_sizes_custom`? It's confusing.

Reply:
Thanks edo - I added a new function set_step_sizes_default that calls set_step_sizes_custom but with no arguments passed, which sets the default step sizes